arangod/Replication2/StateMachines/Document/CollectionReader.cpp

```diff
@@ -28,31 +28,66 @@
 #include "StorageEngine/ReplicationIterator.h"
 #include "StorageEngine/TransactionState.h"
 #include "Transaction/StandaloneContext.h"
-#include "Utils/SingleCollectionTransaction.h"
 #include "VocBase/LogicalCollection.h"
+#include "VocBase/vocbase.h"
 
 namespace arangodb::replication2::replicated_state::document {
-CollectionReader::CollectionReader(
-    std::shared_ptr<LogicalCollection> logicalCollection)
-    : _logicalCollection(std::move(logicalCollection)),
-      _ctx(transaction::StandaloneContext::Create(
-          _logicalCollection->vocbase())) {
+
+ReplicationTransaction::ReplicationTransaction(
+    std::shared_ptr<transaction::Context> ctx)
+    : transaction::Methods(std::move(ctx), options()) {}
+
+auto ReplicationTransaction::options() -> transaction::Options {
   transaction::Options options;
   options.requiresReplication = false;
+  options.fillBlockCache = false;  // TODO - check whether this is a good idea
+  return options;
+}
+
+auto ReplicationTransaction::addCollection(LogicalCollection const& collection)
+    -> Result {
+  return transaction::Methods::addCollection(collection.id(), collection.name(),
```

Comment:
I had a quick look at the Jenkins failure, and this seems to be the culprit. Unfortunately, we cannot call `addCollection` after the transaction has status `CREATED`, which happens after it is constructed. We might have to create a transaction for each shard, which clearly would be less elegant than what you were trying to achieve here, or find some other workaround.

Reply:
We definitely should be able to add readonly collections to an already startet trx. Will look into this!

Reply:
When we try to do that, an exception is thrown in `Methods.cpp:3001`: "cannot add collection to a previously started top-level transaction". But, that can work with `addCollectionAtRuntime`. I don't understand the inner workings of these methods either, just tried it out and it works.